src/tinuous/base.py

```diff
@@ -161,3 +162,32 @@ class BuildLog(BuildAsset):
 
 class Artifact(BuildAsset):
     pass
+
+
+# These config-related classes need to go in this file to avoid a circular
+# import issue:
+
+
+class NoExtraModel(BaseModel):
+    class Config:
+        allow_population_by_field_name = True
+        extra = "forbid"
+
+
+class WorkflowSpec(NoExtraModel):
+    regex: bool = False
+    include: List[Pattern] = Field(default_factory=lambda: [re.compile(".*")])
+    exclude: List[Pattern] = Field(default_factory=list)
+
+    @validator("include", "exclude", pre=True, each_item=True)
+    def _maybe_regex(
+        cls, v: Union[str, Pattern], values: Dict[str, Any]  # noqa: B902, U100
+    ) -> Union[str, Pattern]:
+        if not values["regex"] and isinstance(v, str):
+            v = r"\A" + re.escape(v) + r"\Z"
+        return v
```

Comment:
if I read code right -- you are converting non-regex "workflow" names to "regexes" if regex=False.  Why did you decide to go this way instead of just an explicit conditional within `match` to do regex matching if regex=True and otherwise just simple equality check?

Reply:
I want the regexes to be stored in compiled form for efficiency purposes, and allowing a `WorkflowSpec` to store either strings or compiled patterns would lead to complicated typing situations.

Reply:
FWIW re "for efficiency purposes" -- `re` caches last 100 or so regexes so pragmatically there is no gain to "manually" pre-compile them AFAIK.  If anything, using regex whenever not needed must be less efficient than just a string matching.

Easing typing - ok, makes sense.

Please add a comment in the code so later on me or someone else who stumbles upon it understands at least the motivation.

Reply:
Comment added.